runtime/lua/vim/lsp/rpc.lua

```diff
@@ -608,6 +619,83 @@ function M.connect(host, port)
   end
 end
 
+---@alias vim.lsp.rpc.Dispatcher fun(method: string, params: table<string, any>)
+---@alias vim.lsp.rpc.on_error fun(code: integer, ...: any)
+---@alias vim.lsp.rpc.on_exit fun(code: integer, signal: integer)
+
+---@class vim.lsp.rpc.PrivateDispatchers
+---@field notification vim.lsp.rpc.Dispatcher
+---@field server_request vim.lsp.rpc.Dispatcher
+---@field on_exit vim.lsp.rpc.on_error
+---@field on_error vim.lsp.rpc.on_exit
+
+-- TODO: better name
+---@class vim.lsp.rpc.StartServer
+---@field request fun(method: string, params: table?, callbackfun: fun(err: lsp.ResponseError | nil, result: any), notify_reply_callback:function?)
+---@field notify fun(method: string, params: any)
+---@field is_closing fun(): boolean
+---@field terminate fun(): nil
+
+--- Create a LSP RPC client factory that connects via named pipes (Windows)
+--- or unix domain sockets (Unix) to the given pipe_path (file path on
+--- Unix and name on Windows)
+---
+---@param pipe_path string file path of the domain socket (Unix) or name of the named pipe (Windows) to connect to
+---@return fun(dispatchers: vim.lsp.rpc.PrivateDispatchers): vim.lsp.rpc.StartServer? #function intended to be passed to |vim.lsp.rpc.start_client| or |vim.lsp.start| on the field cmd
+function M.domain_socket_connect(pipe_path)
```

Comment:
Maybe something like:

```suggestion
function M.connect_to_pipe(pipe_path)
```

Merging it into `connect` could also be an option, if we change the args to `opts` and distinguish based on the keys. With a `type(opts) == "table"` check we could even make it backward compatible (if it is a string, it's a host name)

cc @justinmk do you have inputs on this?